routes/project/discount_code.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class Clover
+  hash_branch(:project_prefix, "discount-code") do |r|
+    r.web do
+      authorize("Project:billing", @project.id)
+      billing_path = "#{@project.path}/billing"
+
+      r.post true do
+        discount_code = r.params["discount_code"].to_s.strip.downcase
+        Validation.validate_short_text(discount_code, "discount_code")
+
+        # Check if the discount code exists
+        discount = DiscountCode.where(Sequel.function(:lower, :discount_code) => discount_code).first
+        unless discount
+          flash["error"] = "Discount code not found."
+          r.redirect billing_path
+        end
+
+        # Check if the discount code is expired
+        if discount.expired?
+          flash["error"] = "Discount code has expired."
+          r.redirect billing_path
+        end
```

Comment:
IMHO, we don't need this branch. We can select valid codes using `.where { expires_at > Time.now }`. There's no need to expose too much information, and having fewer branches makes testing easier.